File: scala-server/app/com/bolour/eagerwords/server/domain/CrossWordFinder.scala
/*
 * Copyright 2017-2020 Azad Bolour
 * Licensed under GNU Affero General Public License v3.0 -
 *    https://github.com/azadbolour/eagerwords.com/blob/master/LICENSE.md
 */

package com.bolour.eagerwords.server.domain

import com.bolour.plane.domain.Axis.Axis
import com.bolour.eagerwords.server.util.WordUtil
import com.bolour.eagerwords.common.domain.{PiecePoint, PlayPiece, PlayPieceObj}
import com.bolour.plane.domain.{Axis, Point}
import com.bolour.util.CommonUtil

class CrossWordFinder(board: Board) {
  val dimension = board.dimension
  val grid = board.grid

  // TODO. Move direction constants to a utility object.
  // TODO. Use direction constants.
  val ForwardDir = 1
  val BackwardDir = -1

  /**
    * A crossing is cross word that includes one of the new
    * letters played on a strip.
    *
    * All crossings must be legitimate words in the dictionary.
    */
  def findStripCrossWords(strip: Strip, word: String): List[String] = {
    val l = word.length
    val range = (0 until l).toList
    val crossingIndices = range.filter { i => WordUtil.isBlankChar(strip.content(i)) }
    val acrossWordList = crossingIndices.map{ i =>
      val point = strip.point(i)
      val playedChar = word(i)
      findCrossingWord(point, playedChar, Axis.crossAxis(strip.axis))
    }
    CommonUtil.catOptions(acrossWordList)
  }

  /**
    * Not used for now but is needed when scores of cross plays figure in the total score.
    */
  def findCrossPlays(playPieces: List[PlayPiece]): List[List[(Char, Point, Boolean)]] = {
    val strip = board.stripOfPlay(playPieces)
    val word = PlayPieceObj.playPiecesToWord(playPieces)
    findCrossPlays(strip, word)
  }

  def findCrossPlays(strip: Strip, word: String): List[List[(Char, Point, Boolean)]] = {
    val l = word.length
    val range = (0 until l).toList
    val crossingIndices = range.filter { i => WordUtil.isBlankChar(strip.content(i)) }
    val crossPlayOpts = crossingIndices map { i =>
      val point = strip.point(i)
      crossingPlay(point, word(i), Axis.crossAxis(strip.axis))
    }
    CommonUtil.catOptions(crossPlayOpts)
  }

  def findCrossingWord(crossPoint: Point, crossingChar: Char, axis: Axis): Option[String] = {
    val optCrossPlay: Option[List[(Char, Point, Boolean)]] = crossingPlay(crossPoint, crossingChar, axis)
    optCrossPlay map { play => (play map { case (char, _, _) => char } ).mkString}
    // val word = (play map { case (char, _, _) => char } ).mkString
    // word
  }

  // TODO. Too much going on within this function. Make it more readable as in Haskell server.
  /**
    * Get information about a particular crossword created as a result of
    * playing a letter on a point.
    * @param crossPoint The point at which the letter is played.
    * @param crossingChar The letter played at that point.
    * @param crossAxis The cross axis along which the crossword lies.
    *
    * @return List of tuples (char, point, moved) for the crossword
    *         providing information about the crossword. Each tuple
    *         includes a letter of the crossword, its location on the board,
    *         and whether the letter is being played. The only letter that
    *         is being moved is the one at the cross point. All others
    *         exist on the board. Return None is there is no cross word at the given point.
    */
  def crossingPlay(crossPoint: Point, crossingChar: Char, crossAxis: Axis): Option[List[(Char, Point, Boolean)]] = {
    def toMoveInfo(piecePoint: PiecePoint) = piecePoint match {
      case PiecePoint(piece, point) => (piece.value, point, false)
    }

    val forthNeighborsInfo = board.lineNeighbors(crossPoint, crossAxis, Axis.forward) map toMoveInfo
    val backNeighborsInfo = board.lineNeighbors(crossPoint, crossAxis, Axis.backward) map toMoveInfo

    val crossingMoveInfo = (crossingChar, crossPoint, true)

    if (forthNeighborsInfo.isEmpty && backNeighborsInfo.isEmpty)
      return None
    Some (backNeighborsInfo ++ List(crossingMoveInfo) ++ forthNeighborsInfo)
  }

}

File: haskell-server/src/EagerWords/Server/Domain/CrossWordFinder.hs
--
-- Copyright 2017-2020 Azad Bolour
-- Licensed under GNU Affero General Public License v3.0 -
--   https://github.com/azadbolour/eagerwords/blob/master/LICENSE.md
--

{-# LANGUAGE NamedFieldPuns #-}
{-# LANGUAGE DisambiguateRecordFields #-}
{-# LANGUAGE RecordWildCards #-}
{-# LANGUAGE FlexibleContexts #-}
{-# LANGUAGE ScopedTypeVariables #-}

module EagerWords.Server.Domain.CrossWordFinder (
    findStripCrossWords
  , findCrossPlays
  , findCrossPlay
  ) where

import Data.Maybe (fromJust, isNothing, catMaybes)
import Data.List (transpose, find)

import qualified Bolour.Language.Util.WordUtil as WordUtil
import qualified Bolour.Plane.Domain.Axis as Axis
import qualified Bolour.Plane.Domain.Point as Point
import Bolour.Plane.Domain.Axis (Axis)
import Bolour.Plane.Domain.Point (Point, Point(Point))
import qualified EagerWords.Common.Domain.PiecePoint
import EagerWords.Common.Domain.PiecePoint (PiecePoint, PiecePoint(PiecePoint))
import qualified EagerWords.Common.Domain.Piece as Piece
import EagerWords.Common.Domain.PlayPiece (PlayPiece, MoveInfo)
import qualified EagerWords.Common.Domain.PlayPiece as PlayPiece
import EagerWords.Server.Domain.Board (Board)
import qualified EagerWords.Server.Domain.Board as Board
import EagerWords.Server.Domain.Strip (Strip, Strip(Strip))
import qualified EagerWords.Server.Domain.Strip as Strip
import qualified Data.ByteString.Char8 as BS

forward = Axis.forward
backward = Axis.backward

findStripCrossWords :: Board -> Strip -> String -> [String]
findStripCrossWords board (strip @ Strip {axis, content}) word =
  let range = [0 .. length word - 1]
      crossingIndices = filter (\i -> WordUtil.isBlankChar $ content !! i) range
      calcCrossing :: Int -> Maybe String = \i ->
        let point = Strip.pointAtOffset strip i
            playedChar = word !! i
        in findSurroundingWord board point playedChar (Axis.crossAxis axis)
      crossingStrings = catMaybes (calcCrossing <$> crossingIndices)
      crossingWords = filter (\w -> length w > 1) crossingStrings
  in crossingWords

findSurroundingWord :: Board -> Point -> Char -> Axis -> Maybe String
findSurroundingWord board point letter axis =
  let play = findCrossPlay board point letter axis
  in ((\(char, _, _) -> char) <$>) <$> play

-- | Find the surrounding cross play to a given move (provided as
--   the point and movingLetter parameters).
--
--   Note that the only moving piece in a cross play is the one
--   at the given crossing point.
--   Note also that the moving piece has yet to be placed on the board.
findCrossPlay :: Board -> Point -> Char -> Axis -> Maybe [MoveInfo]
findCrossPlay board point movingLetter crossAxis =
  let crossingMoveInfo = (movingLetter, point, True)
      forthNeighbors = Board.lineNeighbors board point crossAxis Axis.forward
      backNeighbors = Board.lineNeighbors board point crossAxis Axis.backward
      moveInfo neighbor =
        let PiecePoint {piece, point} = neighbor
        in (Piece.value piece, point, False)
  in if null backNeighbors && null forthNeighbors then Nothing
     else Just $ (moveInfo <$> backNeighbors) ++ [crossingMoveInfo] ++ (moveInfo <$> forthNeighbors)

-- | Not used for now but is needed when scores of cross plays figure in the total score.
findCrossPlays :: Board -> [PlayPiece] -> [[MoveInfo]]
findCrossPlays board playPieces =
  let -- TODO. Internal error if fromJust fails.
      strip = fromJust $ Board.stripOfPlay board playPieces
      word = PlayPiece.playPiecesToWord playPieces
  in findCrossPlays' board strip word

findCrossPlays' :: Board -> Strip -> String -> [[MoveInfo]]
findCrossPlays' board (strip @ Strip {axis, content}) word =
  let range = [0 .. length word - 1]
      crossingIndices = filter (\i -> WordUtil.isBlankChar $ content !! i) range
      calcCrossing :: Int -> Maybe [MoveInfo] = \i ->
        let point = Strip.pointAtOffset strip i
            playedChar = word !! i
        in findCrossPlay board point playedChar (Axis.crossAxis axis)
      crossingPlays = calcCrossing <$> crossingIndices
  in catMaybes crossingPlays



